Use backup location in the backup controller

nrb opened the pull request:
Signed-off-by: Nolan Brubaker <nolan@heptio.com>

skriss commented on pkg/controller/backup_controller.go:
```diff
@@ -80,30 +79,28 @@ func NewBackupController(
+		logLevel:             logLevel,
+		pluginRegistry:       pluginRegistry,
+		backupTracker:        backupTracker,
+		backupLocationLister: backupLocationInformer.Lister(),
```
you'll need to wait for the `backupLocationInformer` to sync in the `Run()` func like is being done for the `backupInformer`

nrb replied:
Is this just using a `Synced` field and updating the condition on L165 to look for it? Or do I need a separate condition?

skriss replied:
yep add a `backupLocationListerSynced` field to the struct, set it in the constructor, then add it as an arg to the `WaitForCacheSync(...)` call (it's variadic)

skriss commented on pkg/controller/backup_controller.go:
```diff
@@ -80,30 +79,28 @@ func NewBackupController(
-		pluginRegistry:    pluginRegistry,
-		backupTracker:     backupTracker,
-		metrics:           metrics,
+		backupper:            backupper,
```
it would be nice to convert the controllers to use the `genericController` type while we're in there modifying them.  This requires:

- embedding a `*genericController` in the `backupController`
- initializing it within the `NewBackupController` func using the `newGenericController` func
- removing the `syncHandler` field from the `backupController` (`genericController` has its own) 
- setting the `cacheSyncWaiters` field in the constructor (removes the need to store each of them as their own fields)
- removing the `queue` field
- removing the `Run`, `runWorker`, and `processNextWorkItem` funcs

I did the same for the download request controller -- see https://github.com/heptio/ark/pull/769/files#diff-cdd2d4325fc0100b6d5a4039fc99792e -- although I think I missed at least one step now that I'm testing.

skriss replied:
If you want to defer this to a separate PR I'm ok with that, but let's get it done for 0.10

nrb replied:
I hadn't looked at it yet; want me to push the other fixes then see where we're at?

skriss replied:
sure

skriss commented on pkg/controller/backup_controller.go:
```diff
@@ -382,7 +384,12 @@ func (controller *backupController) runBackup(backup *api.Backup, bucket string)
 	}
 
-	objectStore, err := getObjectStore(controller.objectStoreConfig, pluginManager)
+	backupLocation, err := controller.backupLocationLister.BackupStorageLocations(backup.Namespace).Get(backup.Spec.StorageLocation)
```
I would just pass `backupLocation` into this func in place of `bucket`, then you won't have to fetch it again

skriss commented on pkg/controller/backup_controller.go:
```diff
@@ -458,6 +465,24 @@ func getObjectStore(cloudConfig api.CloudProviderConfig, manager plugin.Manager)
 	return objectStore, nil
 }
 
+// TODO(nrb): Consolidate with other implementations
```
@nrb @carlisia let's say that whichever PR of our 3 merges first will include a function with a new name, like @nrb has here, for getting an object store from a location.  The other 2 of us will then rebase and use that function.  The last person to merge can drop the old function, and rename the new one if they want (or not).  Work for everyone?

nrb replied:
That makes sense to me.

carlisia replied:
Works for me!

skriss commented:
Just noticed that you'll need to add defaulting as well - if no location is specified on the API obj, it should default to the server's `defaultBackupLocation` value

nrb commented:
> Just noticed that you'll need to add defaulting as well

Good catch, will add it + a test case.

nrb commented:
Got default location setting working, but I think I'll need to improve error handling when the location isn't found. Probably make the backup failed.

> ERRO[0095] Error in syncHandler, re-adding item to queue  error="backupstoragelocation.ark.heptio.com \"default\" not found" error.file="backup_controller.go:297" error.function="(*backupController).processBackup" key=heptio-ark/b1 logSource="pkg/controller/backup_controller.go:212"

skriss commented:
Nice - yeah, would be ideal if that could be checked during validation and it's moved to `FailedValidation` if the location doesn't exist

nrb commented:
I've updated to set the default location if the field is empty on the backup, and added some tests around missing locations.

I'm not super happy with how the tests are implemented, so open to suggestions there.

skriss commented on pkg/controller/backup_controller_test.go:
```diff
@@ -224,6 +244,43 @@ func TestProcessBackup(t *testing.T) {
 					mock.Anything, // actions
 				).Return(nil)
 
+				var location *v1.BackupStorageLocation
```
What I would do here is always add two locations to the informer - `default` and `loc1`.  Then, keep the 3 test cases you have, but change the third to use location `loc2` (which won't exist).  Your verification can stay as-is but the setup will be more straightforward

skriss commented:
The rest LGTM - will take a final look Monday but should be good to merge

nrb commented:
Done. I think it will be even cleaner with the `newBackupLocation` function, but no real need to hold this up just for that, either.
